pkg/lib/ncl/envelope_serde.go

```diff
@@ -0,0 +1,112 @@
+package ncl
+
+import (
+	"encoding/binary"
+	"hash/crc32"
+)
+
+// Error message constants
+const (
+	ErrMsgTooShort  = "too short"
+	ErrMsgCRCFailed = "CRC check failed"
+)
+
+const (
+	VersionSize = 1
+	CRCSize     = 4
+	HeaderSize  = VersionSize + CRCSize
+)
+
+// EnvelopeSerializer handles the serialization and deserialization of messages
+// with version information and CRC checks. It wraps the actual message serialization
+// with additional metadata for versioning and integrity checking.
+//
+// Envelope Structure:
+// +----------------+----------------+--------------------+
+// | Version (1 byte)| CRC (4 bytes) | Serialized Message |
+// +----------------+----------------+--------------------+
+//
+// - Version: Indicates the schema version used for serialization (1 byte)
+// - CRC: A 32-bit CRC checksum of the serialized message (4 bytes)
+// - Serialized Message: The actual message content, serialized by a version-specific serializer
+//
+// The EnvelopeSerializer adds a version byte and a CRC checksum to each serialized message,
+// allowing for future extensibility, backward compatibility, and data integrity verification.
+type EnvelopeSerializer struct {
+	serializerSchemaVersion SchemaVersion
```

Comment:
This isn't much clear, in my opinion. Your envelope serializer has a schema version, but you store multiple serializers for different schema versions too ! It is bit confusing. What do you think ?

Reply:
Yeah it wasn't clear. The `serializerSchemaVersion` dictates how new messages are serialized. We store multiple serializers to help with deserializing messages that were encoded differently for backward compatibility. I have renamed the field and added more comments in https://github.com/bacalhau-project/bacalhau/pull/4257/commits/7bbe03aa460f969dcaa842dd627fc934c7340bef